File: src/core/vultr-manager.ts
/**
 * Read-safe Vultr API v2 client — the observable half of the integration.
 *
 * Like `git-manager.ts` and `cloudflare-manager.ts`, this module deliberately
 * contains no operation that changes a machine. Start / halt / reboot live in
 * `vultr-actions.ts`, and destructive operations (delete, reinstall, restore)
 * are not implemented at all — the same line `GitManager` draws around
 * `reset --hard`.
 *
 * Every field and path below comes from Vultr's official Go SDK
 * (`github.com/vultr/govultr`), which is generated against the same API the
 * docs describe.
 */

import type { ProviderFetch } from "./providers/egress.js";

/** Vultr's `status` — the account-level lifecycle of the machine. */
export type VultrInstanceStatus = "active" | "pending" | "suspended" | "resizing";

export interface VultrAccount {
  name: string;
  email: string;
  /** What this API key is permitted to do — `manage_users`, `subscriptions`, … */
  acls: string[];
}

export interface VultrInstance {
  id: string;
  label: string;
  hostname: string;
  status: VultrInstanceStatus;
  /** `running` / `stopped`. Empty until the instance finishes provisioning. */
  powerStatus: string;
  /** `none` / `locked` / `installingbooting` / `ok`. */
  serverStatus: string;
  /** Empty string while the instance has no address yet, normalized to undefined. */
  mainIp?: string;
  v6MainIp?: string;
  internalIp?: string;
  region: string;
  plan: string;
  os: string;
  vcpuCount: number;
  /** Vultr reports RAM in MB and disk in GB. */
  ramMb: number;
  diskGb: number;
  tags: string[];
  createdAt?: number;
  /**
   * Which login the instance was built with: `root`, or `limited` for the
   * unprivileged `linuxuser` account Vultr creates instead. This is the one
   * field that decides whether an adopted SSH target can actually connect.
   */
  userScheme: string;
}

export class VultrApiError extends Error {
  constructor(
    message: string,
    public readonly status: number,
    public readonly path: string,
  ) {
    super(message);
    this.name = "VultrApiError";
  }
}

/** Vultr caps `per_page` at 500; one page covers any realistic account. */
const PER_PAGE = 500;
/** A hard stop on cursor following, so a pathological account cannot hang a request. */
const MAX_PAGES = 10;

export class VultrManager {
  constructor(
    private readonly token: string,
    private readonly baseUrl = "https://api.vultr.com/v2",
    /** The provider's scoped `fetch`; global `fetch` when built outside the registry. */
    private readonly fetchImpl?: ProviderFetch,
  ) {}

  /**
   * The account this API key belongs to.
   *
   * Vultr gives an account no id — the email is what identifies one — so that
   * is what the neutral `ProviderAccount.id` carries.
   */
  async account(): Promise<VultrAccount> {
    const { account } = await this.request<{ account: RawAccount }>("/account");
    return {
      name: account.name ?? "",
      email: account.email ?? "",
      acls: account.acls ?? [],
    };
  }

  /**
   * Every instance on the account, following Vultr's cursor pagination.
   *
   * The whole list rather than a page: the caller is building an SSH server
   * list, and a partial one would silently hide machines.
   */
  async listInstances(opts: { label?: string; tag?: string; region?: string } = {}): Promise<
    VultrInstance[]
  > {
    const instances: VultrInstance[] = [];
    let cursor: string | undefined;
    for (let page = 0; page < MAX_PAGES; page += 1) {
      const query = new URLSearchParams({ per_page: String(PER_PAGE) });
      if (opts.label) query.set("label", opts.label);
      if (opts.tag) query.set("tag", opts.tag);
      if (opts.region) query.set("region", opts.region);
      if (cursor) query.set("cursor", cursor);

      const body = await this.request<RawInstanceList>(`/instances?${query.toString()}`);
      instances.push(...(body.instances ?? []).map(normalizeInstance));
      cursor = body.meta?.links?.next || undefined;
      if (!cursor) break;
    }
    return instances;
  }

  async getInstance(id: string): Promise<VultrInstance> {
    const { instance } = await this.request<{ instance: RawInstance }>(
      `/instances/${encodeURIComponent(id)}`,
    );
    return normalizeInstance(instance);
  }

  private request<T>(path: string, opts?: { method?: string; body?: unknown }): Promise<T> {
    return vultrRequest<T>(
      { token: this.token, baseUrl: this.baseUrl, fetch: this.fetchImpl },
      path,
      opts,
    );
  }
}

export interface VultrRequestAuth {
  token: string;
  baseUrl?: string;
  /**
   * The provider's scoped `fetch` (see `providers/egress.ts`). Absent means
   * global `fetch` — the case for a manager built directly in a test. Production
   * clients come from `vultr-context.ts`, which always supplies one.
   */
  fetch?: ProviderFetch;
}

/**
 * The one place a Vultr API call is made. Shared with `vultr-actions.ts` so
 * auth and error shaping stay identical across the read and write halves — the
 * read/write boundary is which module exposes an operation, not which one can
 * reach the network.
 */
export async function vultrRequest<T>(
  auth: VultrRequestAuth,
  path: string,
  opts?: { method?: string; body?: unknown },
): Promise<T> {
  const url = path.startsWith("http") ? path : `${auth.baseUrl ?? "https://api.vultr.com/v2"}${path}`;
  const headers: Record<string, string> = {
    Authorization: `Bearer ${auth.token}`,
    Accept: "application/json",
  };
  const init: RequestInit = { method: opts?.method ?? "GET", headers };
  if (opts?.body !== undefined) {
    headers["Content-Type"] = "application/json";
    init.body = JSON.stringify(opts.body);
  }

  const response = await (auth.fetch ?? fetch)(url, init);
  // Power actions answer 204 with no body, so the body is only read when there
  // is one to read.
  const text = await response.text().catch(() => "");
  if (!response.ok) {
    const failure = safeParse(text) as { error?: string } | undefined;
    throw new VultrApiError(
      failure?.error || `Vultr returned ${response.status} for ${path}`,
      response.status,
      path,
    );
  }
  return (safeParse(text) ?? {}) as T;
}

function safeParse(text: string): unknown {
  if (!text.trim()) return undefined;
  try {
    return JSON.parse(text);
  } catch {
    return undefined;
  }
}

interface RawAccount {
  name?: string;
  email?: string;
  acls?: string[];
}

interface RawInstanceList {
  instances?: RawInstance[];
  meta?: { total?: number; links?: { next?: string; prev?: string } | null } | null;
}

interface RawInstance {
  id: string;
  label?: string;
  hostname?: string;
  status?: string;
  power_status?: string;
  server_status?: string;
  main_ip?: string;
  v6_main_ip?: string;
  internal_ip?: string;
  region?: string;
  plan?: string;
  os?: string;
  vcpu_count?: number;
  ram?: number;
  disk?: number;
  tags?: string[];
  date_created?: string;
  user_scheme?: string;
}

/**
 * Vultr writes "no address" as the empty string rather than omitting the field,
 * and `0.0.0.0` while an instance is still being built. Both mean the same
 * thing to a caller trying to reach the machine.
 */
function address(value: string | undefined): string | undefined {
  const trimmed = value?.trim();
  if (!trimmed || trimmed === "0.0.0.0" || trimmed === "::") return undefined;
  return trimmed;
}

function normalizeInstance(instance: RawInstance): VultrInstance {
  const createdAt = Date.parse(instance.date_created ?? "");
  return {
    id: instance.id,
    label: instance.label ?? "",
    hostname: instance.hostname ?? "",
    status: (instance.status as VultrInstanceStatus) ?? "pending",
    powerStatus: instance.power_status ?? "",
    serverStatus: instance.server_status ?? "",
    mainIp: address(instance.main_ip),
    v6MainIp: address(instance.v6_main_ip),
    internalIp: address(instance.internal_ip),
    region: instance.region ?? "",
    plan: instance.plan ?? "",
    os: instance.os ?? "",
    vcpuCount: instance.vcpu_count ?? 0,
    ramMb: instance.ram ?? 0,
    diskGb: instance.disk ?? 0,
    tags: instance.tags ?? [],
    createdAt: Number.isNaN(createdAt) ? undefined : createdAt,
    userScheme: instance.user_scheme ?? "root",
  };
}

File: crates/nomoreide-core/src/vultr_manager.rs
//! The one place a Vultr API call is made.
//!
//! The Rust half of `src/core/vultr-manager.ts`. Shared with the actions crate
//! so auth and error shaping stay identical across the read and write halves —
//! the read/write boundary is which module exposes an operation, not which one
//! can reach the network.

use crate::providers::api_base::{provider_api_base, provider_api_host};
use crate::providers::egress::ProviderEgress;
use serde_json::Value;

const API_BASE: &str = "https://api.vultr.com/v2";

/// Vultr's own maximum, so an account of any size is one request per page.
const PAGE_SIZE: u32 = 500;

/// A hard stop on cursor following, so a pathological account cannot hang a
/// request.
const MAX_PAGES: usize = 10;

pub fn api_base() -> String {
    provider_api_base("NOMOREIDE_VULTR_API_BASE", API_BASE)
}

fn egress() -> ProviderEgress {
    ProviderEgress::new("vultr", vec![provider_api_host(&api_base())])
}

#[derive(Debug, Clone)]
pub struct VultrApiError {
    pub message: String,
    pub status: u16,
}

impl std::fmt::Display for VultrApiError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}", self.message)
    }
}

impl std::error::Error for VultrApiError {}

/// One request, through the provider's egress boundary.
pub async fn request(
    token: &str,
    path: &str,
    method: reqwest::Method,
) -> Result<Value, VultrApiError> {
    let url = format!("{}{path}", api_base());
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::none())
        .build()
        .map_err(|error| VultrApiError {
            message: format!("Vultr request failed: {error}"),
            status: 0,
        })?;

    let bearer = token.to_string();
    let response = egress()
        .send(
            &client,
            &url,
            |client, verb, target| {
                client
                    .request(verb, target)
                    .header("Authorization", format!("Bearer {bearer}"))
                    .header("Accept", "application/json")
            },
            method,
        )
        .await
        .map_err(|error| VultrApiError {
            message: format!("Vultr request failed: {error}"),
            status: 0,
        })?;

    let status = response.status();
    let text = response.text().await.unwrap_or_default();
    let body: Value = serde_json::from_str(&text).unwrap_or(Value::Null);
    if status.is_success() {
        return Ok(body);
    }
    Err(VultrApiError {
        message: api_error_message(&body, status.as_u16(), path),
        status: status.as_u16(),
    })
}

/// Vultr puts the useful sentence in a top-level `error`. When the body is not
/// JSON at all — a proxy's HTML error page, say — the bare status says nothing
/// about what was being asked, so the request is named instead.
fn api_error_message(body: &Value, status: u16, path: &str) -> String {
    body.get("error")
        .and_then(Value::as_str)
        .filter(|message| !message.is_empty())
        .map(str::to_string)
        .unwrap_or_else(|| format!("Vultr returned {status} for {path}"))
}

/// One path segment, encoded the way `encodeURIComponent` encodes it.
fn segment(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

pub struct VultrManager {
    token: String,
}

impl VultrManager {
    pub fn new(token: impl Into<String>) -> Self {
        Self {
            token: token.into(),
        }
    }

    /// The signed-in account.
    pub async fn account(&self) -> Result<Value, VultrApiError> {
        let body = request(&self.token, "/account", reqwest::Method::GET).await?;
        Ok(body.get("account").cloned().unwrap_or(Value::Null))
    }

    /// Every instance, following the cursor Vultr hands back.
    ///
    /// The walk stops on an empty `next` — the vendor's own end-of-list signal —
    /// or at `MAX_PAGES`, whichever comes first.
    pub async fn list_instances(&self) -> Result<Vec<Value>, VultrApiError> {
        let mut instances = Vec::new();
        let mut cursor: Option<String> = None;

        for _ in 0..MAX_PAGES {
            let path = match &cursor {
                None => format!("/instances?per_page={PAGE_SIZE}"),
                Some(next) => format!("/instances?per_page={PAGE_SIZE}&cursor={}", segment(next)),
            };
            let body = request(&self.token, &path, reqwest::Method::GET).await?;
            if let Some(page) = body.get("instances").and_then(Value::as_array) {
                instances.extend(page.iter().cloned());
            }
            let next = body
                .get("meta")
                .and_then(|meta| meta.get("links"))
                .and_then(|links| links.get("next"))
                .and_then(Value::as_str)
                .unwrap_or_default()
                .to_string();
            if next.is_empty() {
                break;
            }
            cursor = Some(next);
        }
        Ok(instances)
    }

    /// One instance.
    pub async fn instance(&self, id: &str) -> Result<Value, VultrApiError> {
        let path = format!("/instances/{}", segment(id));
        let body = request(&self.token, &path, reqwest::Method::GET).await?;
        Ok(body.get("instance").cloned().unwrap_or(Value::Null))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn the_vendors_sentence_wins_over_the_status() {
        let body = json!({ "error": "Instance not found", "status": 404 });
        assert_eq!(
            api_error_message(&body, 404, "/instances/x"),
            "Instance not found"
        );
    }

    #[test]
    fn a_body_that_is_not_json_names_the_request() {
        assert_eq!(
            api_error_message(&Value::Null, 500, "/instances/x"),
            "Vultr returned 500 for /instances/x"
        );
    }

    #[test]
    fn an_id_with_a_space_is_encoded() {
        assert_eq!(segment("inst space"), "inst%20space");
    }
}
